include/nbl/builtin/hlsl/workgroup2/shared_scan.hlsl

```diff
@@ -0,0 +1,419 @@
+                value[0] = binop(left, left_last_elem);
+            }
+            else
+            {
+                [unroll]
+                for (uint32_t i = 0; i < Config::ItemsPerInvocation_0; i++)
+                    value[i] = binop(left, value[i]);
+            }
+            dataAccessor.template set<vector_lv0_t>(idx * Config::WorkgroupSize + virtualInvocationIndex, value);
+        }
+    }
+};
+
+// 3-level scans
+template<class Config, class BinOp, class device_capabilities>
+struct reduce<Config, BinOp, 3, device_capabilities>
+{
+    using scalar_t = typename BinOp::type_t;
+    using vector_lv0_t = vector<scalar_t, Config::ItemsPerInvocation_0>;   // data accessor needs to be this type
+    using vector_lv1_t = vector<scalar_t, Config::ItemsPerInvocation_1>;
+    using vector_lv2_t = vector<scalar_t, Config::ItemsPerInvocation_2>;
+
+    template<class DataAccessor, class ScratchAccessor>
+    scalar_t __call(NBL_REF_ARG(DataAccessor) dataAccessor, NBL_REF_ARG(ScratchAccessor) scratchAccessor)
+    {
+        using config_t = subgroup2::Configuration<Config::SubgroupSizeLog2>;
+        using params_lv0_t = subgroup2::ArithmeticParams<config_t, BinOp, Config::ItemsPerInvocation_0, device_capabilities>;
+        using params_lv1_t = subgroup2::ArithmeticParams<config_t, BinOp, Config::ItemsPerInvocation_1, device_capabilities>;
+        using params_lv2_t = subgroup2::ArithmeticParams<config_t, BinOp, Config::ItemsPerInvocation_2, device_capabilities>;
+        BinOp binop;
+
+        const uint32_t invocationIndex = workgroup::SubgroupContiguousIndex();
+        // level 0 scan
+        subgroup2::reduction<params_lv0_t> reduction0;
+        [unroll]
+        for (uint32_t idx = 0, virtualInvocationIndex = invocationIndex; idx < Config::VirtualWorkgroupSize / Config::WorkgroupSize; idx++)
+        {
+            vector_lv0_t scan_local;
+            dataAccessor.template get<vector_lv0_t>(idx * Config::WorkgroupSize + virtualInvocationIndex, scan_local);
+            scan_local = reduction0(scan_local);
+            if (glsl::gl_SubgroupInvocationID()==Config::SubgroupSize-1)
+            {
+                const uint32_t virtualSubgroupID = Config::virtualSubgroupID(glsl::gl_SubgroupID(), idx);
+                const uint32_t bankedIndex = Config::sharedMemCoalescedIndex(virtualSubgroupID, Config::ItemsPerInvocation_1);    // (virtualSubgroupID & (Config::ItemsPerInvocation_1-1)) * Config::SubgroupsPerVirtualWorkgroup + (virtualSubgroupID/Config::ItemsPerInvocation_1);
+                scratchAccessor.template set<scalar_t>(bankedIndex, scan_local[Config::ItemsPerInvocation_0-1]);   // set last element of subgroup scan (reduction) to level 1 scan
+            }
+        }
+        scratchAccessor.workgroupExecutionAndMemoryBarrier();
+
+        // level 1 scan
+        subgroup2::reduction<params_lv1_t> reduction1;
+        if (glsl::gl_SubgroupID() < Config::SubgroupSizeLog2*Config::ItemsPerInvocation_1)
```

Comment:
huh ?

You have `VirtualWorkgroupSize` invocations producing results, meaning that to digest them you need `VirtualWorkgroupSize/ItemsPerInvocation_1/SubgroupSize == (VirtualWorkgroupSize>>SubgroupSize)/ItemsPerInvocation_1` invocations

and `VirtualWorkgroupSize/SubgroupSize/SubgroupSize /ItemsPerInvocation_1` subgroups, how do you arrive at your inequality here? Alos involving a Log2 !?

I can kinda see that `SubgroupSize*ItemsPerInvocation_2` outputs will need to be produced by level 1 to be consumed by level 2, but then your inequality must be `< SubgroupSize*ItemsPerInvocation_2` or `< (ItemsPerInovcation_2>>SubgroupSizeLog2)`